### ISSUE TITLE
optimize_read_in_order doesn't work with ORDER BY (toDate(timestamp)) in clickhouse 20.8+

### ISSUE THREAD

**Issue by UnamedRus** [ClickHouse/ClickHouse#16107]
**How to reproduce**
```
CREATE TABLE test_order_by (timestamp DateTime, key UInt32) ENGINE=MergeTree() PARTITION BY toYYYYMM(timestamp) ORDER BY (toDate(timestamp), key);
INSERT INTO test_order_by SELECT now() + toIntervalSecond(number), number % 4 FROM numbers(10000000);
OPTIMIZE TABLE test_order_by FINAL;
Q1: SELECT * FROM test_order_by ORDER BY timestamp LIMIT 10;
Q2: SELECT * FROM test_order_by ORDER BY toDate(timestamp) LIMIT 10;
Q3: SELECT * FROM test_order_by ORDER BY toDate(timestamp), timestamp LIMIT 10;

Q1 works the same on all versions of clickhouse.
Q1: 10 rows in set. Elapsed: 0.060 sec. Processed 10.00 million rows, 80.00 MB (166.87 million rows/s., 1.33 GB/s.)

Clickhouse 20.3.19
Q2: 10 rows in set. Elapsed: 0.013 sec. Processed 655.36 thousand rows, 5.24 MB (50.57 million rows/s., 404.54 MB/s.)
Q3: 10 rows in set. Elapsed: 0.034 sec. Processed 720.90 thousand rows, 5.77 MB (21.15 million rows/s., 169.23 MB/s.)
Clickhouse 20.4.2.9
Q2: Code: 15. DB::Exception: Received from localhost:9000. DB::Exception: Column 'toDate(timestamp)' already exists.
Q3: Code: 15. DB::Exception: Received from localhost:9000. DB::Exception: Column 'toDate(timestamp)' already exists.
Clickhouse 20.5.5.74
Q2: 10 rows in set. Elapsed: 0.020 sec. Processed 1.44 million rows, 11.53 MB (72.88 million rows/s., 583.04 MB/s.)
Q3: 10 rows in set. Elapsed: 0.013 sec. Processed 1.57 million rows, 12.58 MB (121.66 million rows/s., 973.31 MB/s.)
Clickhouse 20.6.8.5
Q2: 10 rows in set. Elapsed: 0.008 sec. Processed 1.05 million rows, 8.39 MB (131.13 million rows/s., 1.05 GB/s.)
Q3: 10 rows in set. Elapsed: 0.011 sec. Processed 1.11 million rows, 8.91 MB (105.19 million rows/s., 841.48 MB/s.)
Clickhouse 20.7.4.11
Q2: 10 rows in set. Elapsed: 0.008 sec. Processed 1.05 million rows, 8.38 MB (130.08 million rows/s., 1.04 GB/s.)
Q3: 10 rows in set. Elapsed: 0.012 sec. Processed 1.11 million rows, 8.91 MB (90.35 million rows/s., 722.80 MB/s.)
Clickhouse 20.8.4.11
Q2: 10 rows in set. Elapsed: 0.046 sec. Processed 10.00 million rows, 80.00 MB (215.31 million rows/s., 1.72 GB/s.)
Q3: 10 rows in set. Elapsed: 0.054 sec. Processed 10.00 million rows, 80.00 MB (185.84 million rows/s., 1.49 GB/s.)
Clickhouse 20.9.3
Q2: 10 rows in set. Elapsed: 0.055 sec. Processed 10.00 million rows, 80.00 MB (181.47 million rows/s., 1.45 GB/s.)
Q3: 10 rows in set. Elapsed: 0.040 sec. Processed 10.00 million rows, 80.00 MB (250.70 million rows/s., 2.01 GB/s.)


```
**Additional context**
It would be a good feature, of clickhouse will use optimize in order with monotonic functions without ORDER BY rewriting.
` SET optimize_monotonous_functions_in_order_by = 1` has no affect on that kind of queries.

**Comment by CurtizJ** [comment]
But with `SET optimize_monotonous_functions_in_order_by = 0` `optimize_read_in_order` works :)

I guess something wrong is in rewriting ASTs.

**Comment by CurtizJ** [comment]
It's because `optimize_monotonous_functions_in_order_by` optimization drops `toDate` function as redundant and then we can't match `ORDER BY` expression with sorting key to apply `optimize_monotonous_functions_in_order_by` optimization.
```sql
EXPLAIN SYNTAX
SELECT *
FROM test_order_by
ORDER BY
    toDate(timestamp) ASC,
    timestamp ASC
LIMIT 10

┌─explain────────────┐
│ SELECT             │
│     timestamp,     │
│     key            │
│ FROM test_order_by │
│ ORDER BY           │
│     timestamp ASC, │
│     timestamp ASC  │
│ LIMIT 10           │
└────────────────────┘
```

**Comment by alexey-milovidov** [comment]
@CurtizJ There is an evidence that this is still broken. Example:

```
EXPLAIN SYNTAX
SELECT DISTINCT traceID
FROM jaeger_index
WHERE (service = 'loadbalancer')
AND ((timestamp >= (now() - toIntervalHour(1))) AND (timestamp <= now()))
AND hasAll(tags.hash, [cityHash64('http.status_code', '200')])
ORDER BY
    service ASC,
    -toUnixTimestamp(timestamp) ASC
LIMIT 20

Query id: 2d954deb-cd87-4a6b-b139-cbd67880c09f

┌─explain────────────────────────────────────────────────────────────────────┐
│ SELECT DISTINCT traceID                                                    │
│ FROM jaeger_index                                                          │
│ WHERE (service = 'loadbalancer')                                           |
| AND ((timestamp >= (now() - toIntervalHour(1))) AND (timestamp <= now()))  |
| AND hasAll(tags.hash, [cityHash64('http.status_code', '200')])             |
│ ORDER BY                                                                   │
│     service ASC,                                                           │
│     timestamp DESC                                                         │
│ LIMIT 20                                                                   │
└────────────────────────────────────────────────────────────────────────────┘
```

Compare:

```
SELECT DISTINCT traceID
FROM jaeger_index
WHERE (service = 'loadbalancer') AND ((timestamp >= (now() - toIntervalHour(1))) AND (timestamp <= now())) AND hasAll(tags.hash, [cityHash64('http.status_code', '200')])
ORDER BY
    service ASC,
    -toUnixTimestamp(timestamp) ASC
LIMIT 20

Query id: 731914b1-5067-4b1b-8544-7a1e75599364

┌─traceID──────────────────────────┐
│ 9df385dc84e376bfdb40277c053145df │
│ 943d3380dd3f155f90b189e14b206384 │
│ 707a76fad314fd28acd886b2983087d7 │
│ 86d48cf673828ff88830c8b0996565f6 │
│ 698dad121dd4bece                 │
│ c073888abf3b530389b0ad46ff4070a1 │
│ bc1488c3f5f26f13ec54b1b66db1a654 │
│ 0f383a6860051d5fb887e7fda7a46423 │
│ c3a4ecacbbd95888a7d37debab54b693 │
│ 4a51fa697cad92c565093e5058b135bf │
│ 710315a72ed3bbb44427beb81e306ad1 │
│ 9abc8442f3e60f1c5821eb5a44361eac │
│ 9d82361a066c3956989bb5e95691b4bd │
│ 9b14ecb2c5ba5e66                 │
│ b977e040b548ca93                 │
│ 99ff943b64da8cc043751463e23c1529 │
└──────────────────────────────────┘
┌─traceID──────────────────────────┐
│ 66b2edbf29f2fcc0                 │
│ c9fda2d85ea054a4                 │
│ 370f1e2ce5dc6d7afc13675afc31682c │
│ c389d3d707ae8b4c662e835efda61c93 │
└──────────────────────────────────┘

20 rows in set. Elapsed: 7.565 sec. Processed 135.19 million rows, 29.17 GB (17.87 million rows/s., 3.86 GB/s.)
```

And:

```
SELECT DISTINCT traceID
FROM jaeger_index
WHERE (service = 'loadbalancer') AND ((timestamp >= (now() - toIntervalHour(1))) AND (timestamp <= now())) AND hasAll(tags.hash, [cityHash64('http.status_code', '200')])
ORDER BY
    service ASC,
    -toUnixTimestamp(timestamp) ASC
LIMIT 20
SETTINGS optimize_monotonous_functions_in_order_by = 0

Query id: de3d5a86-675e-4c2f-9f61-b3bb9ba71bfd

┌─traceID──────────────────────────┐
│ 028bd84269d5bd38                 │
│ 69695151177288bc                 │
│ 7dc13a13395c0673c35386cf37ea1419 │
│ 30706d972acef4c8                 │
│ 500d806e96e04cb68663314fd9e3fb4b │
│ 66ad2cf27e86fe00                 │
│ 0dbfb3909178cbe881b2ef2db07a88fa │
│ 6ae0e4db2590b109                 │
│ 393f56980c54a5816818317d57a06cbf │
│ e77a8987cc07b3a14289a388f3f08672 │
│ 4995ede8128d3bf1                 │
│ c13c2f8a1fd8f56e2b6304a4a4c1b03f │
│ 9db5157e7f1f3f0ec262ef46ff49f37a │
│ 6f0359192bf72174                 │
│ a04dac0291bec6824342d9e788f7f36f │
│ 34d8188e3ccc5244                 │
│ 6265587d0b22966d                 │
│ 112f754c7c47650e845854b9d572cae7 │
│ f8b39054647879e2                 │
│ 4d70f9d99dbbf711de56bc0dec2912d1 │
└──────────────────────────────────┘

20 rows in set. Elapsed: 0.106 sec. Processed 5.71 million rows, 1.22 GB (53.95 million rows/s., 11.50 GB/s.)
```

**Comment by alexey-milovidov** [comment]
@CurtizJ said that it is most likely fixed in https://github.com/ClickHouse/ClickHouse/pull/32670 and will be available in 22.1.
